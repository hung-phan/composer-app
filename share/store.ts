import { Middleware, configureStore } from "@reduxjs/toolkit";
import { produce } from "immer";
import { HYDRATE, MakeStore, createWrapper } from "next-redux-wrapper";
import { AnyAction, Store, StoreEnhancer, combineReducers } from "redux";

import {
  selectors,
  mountPoint as templateEngineMountPoint,
  reducer as templateEngineReducer,
} from "./domain/engine/coreEngineStateStore";
import { convertToClass, encode } from "./domain/engine/serializers";
import { Node } from "./domain/interfaces";

export const middlewares: Middleware[] = [];
export const enhancers: StoreEnhancer[] = [];

const applicationReducer = combineReducers({
  [templateEngineMountPoint]: templateEngineReducer,
});

export type RootState = ReturnType<typeof applicationReducer>;

export const makeStore: MakeStore<Store<RootState>> = () =>
  configureStore({
    reducer: (state: RootState, action: AnyAction) =>
      produce(state, (draft) => {
        switch (action.type) {
          case HYDRATE:
            Object.assign(draft, action.payload);
            break;

          default:
            return applicationReducer(draft, action);
        }
      }),
    devTools:
      process.env.ENVIRONMENT === "client" &&
      process.env.NODE_ENV === "development",
    middleware: (getDefaultMiddleware) =>
      getDefaultMiddleware({
        serializableCheck: false,
      }).concat(...middlewares),
    enhancers,
  });

export const wrapper = createWrapper<Store<RootState>>(makeStore, {
  serializeState: encode,
  deserializeState: (encodedData: string) => {
    const state = JSON.parse(encodedData);

    const templateEngineState = selectors.getState(state);

    for (const [key, value] of Object.entries(templateEngineState)) {
      templateEngineState[key] = convertToClass<Node>(value);
    }

    return state;
  },
});
