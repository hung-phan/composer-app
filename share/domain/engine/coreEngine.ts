import { Builder } from "builder-pattern";
import * as _ from "lodash";
import Router from "next/router";
import { AnyAction } from "redux";
import { ThunkDispatch } from "redux-thunk";

import fetch from "../../library/fetch";
import { Id } from "../../library/idGenerator";
import { RootState } from "../../store";
import {
  AddInListElementMethod,
  BatchRenderElementMethod,
  ClientInfo,
  DataContainer,
  DeleteInListElementMethod,
  Element,
  HttpMethod,
  HttpMethodRequestBody,
  InvokeExternalMethod,
  Method,
  NavigateMethod,
  RenderElementMethod,
  Response,
  UpdateElementMethod,
  UpdateInListElementMethod,
  UpdateStateMethod,
} from "../interfaces";
import { getSimplifiedElement } from "./coreEngineHelpers";
import { actions, selectors } from "./coreEngineStateStore";
import getTaskQueue, { TaskQueue } from "./queue";
import { decode } from "./serializers";

class CoreEngine {
  readonly ownerId: Id;
  readonly taskQueue: TaskQueue;
  readonly dispatch: ThunkDispatch<RootState, unknown, AnyAction>;
  readonly clientInfo?: ClientInfo<any>;
  actions: AnyAction[] = [];

  addToDispatchQueue(action: AnyAction) {
    this.actions.push(action);
  }

  flushDispatchQueue() {
    if (_.isEmpty(this.actions)) {
      return;
    }

    this.dispatch(actions.handleCoreEngineActions(this.actions));

    this.actions = [];
  }
}

async function evalMethod(
  method: Method,
  coreEngine: CoreEngine
): Promise<void> {
  if (method === null) {
    return;
  }

  if (method instanceof InvokeExternalMethod) {
    coreEngine.addToDispatchQueue(method.action);
  } else if (method instanceof UpdateStateMethod) {
    coreEngine.addToDispatchQueue(
      actions.updateStateElement({
        stateElementId: method.stateElementId,
        patches: method.patches,
      })
    );
  } else if (method instanceof HttpMethod) {
    if (_.isEmpty(method.stateIds)) {
      await makeHttpCall(method, coreEngine);
    } else {
      await coreEngine.dispatch((__, getState) => {
        const currentState = getState();

        return makeHttpCall(
          method,
          coreEngine,
          method.stateIds!.map((clientStateId) =>
            selectors.getElementState(currentState, clientStateId)
          )
        );
      });
    }
  } else if (method instanceof RenderElementMethod) {
    await registerElement(method.element, coreEngine);
  } else if (method instanceof BatchRenderElementMethod) {
    await Promise.all(
      method.elements.map((element) => registerElement(element, coreEngine))
    );
  } else if (method instanceof UpdateElementMethod) {
    await coreEngine.dispatch(async (__, getState) => {
      const parentElement = selectors.getParentElement(getState(), method.id);

      await registerElement(method.element, coreEngine);

      coreEngine.addToDispatchQueue(
        actions.replaceElement({
          parentId: parentElement.id,
          oldId: method.id,
          id: method.element.id,
        })
      );
    });
  } else if (method instanceof AddInListElementMethod) {
    await Promise.all(
      method.elements.map((element) => registerElement(element, coreEngine))
    );
    coreEngine.addToDispatchQueue(
      actions.addElementInList({
        oldId: method.id,
        ids: method.elements.map((element) => element.id),
      })
    );
  } else if (method instanceof UpdateInListElementMethod) {
    await Promise.all(
      method.elements.map((element) => registerElement(element, coreEngine))
    );
    coreEngine.addToDispatchQueue(
      actions.updateElementInList({
        oldId: method.id,
        ids: method.elements.map((element) => element.id),
      })
    );
  } else if (method instanceof DeleteInListElementMethod) {
    coreEngine.addToDispatchQueue(
      actions.deleteElementInList({ ids: method.ids })
    );
  } else if (method instanceof NavigateMethod) {
    if (process.env.ENVIRONMENT === "client") {
      await Router.push(method.url);
    } else {
      throw new Error(
        `Unsupported method for serverSide: ${method.interfaceName}`
      );
    }
  } else {
    throw new Error(`Unsupported method: ${method.interfaceName}`);
  }
}

async function makeHttpCall(
  method: HttpMethod<any>,
  coreEngine: CoreEngine,
  elementStates?: DataContainer[]
): Promise<void> {
  if (method.requestType === undefined) {
    throw new Error("requestType cannot be undefined");
  }

  if (method.before) {
    await dispatchTask(coreEngine, method.before);
  }

  const requestOptions: RequestInit = {};
  const clientInfo = _.merge(coreEngine.clientInfo, method.forwardedClientInfo);
  const requestBody: HttpMethodRequestBody<any, any> = {
    elementStates,
    clientInfo:
      _.isEmpty(clientInfo) || _.isEmpty(clientInfo.data)
        ? undefined
        : clientInfo,
    requestData: method.requestData,
  };
  let serverUrl = method.url;

  if (method.requestType === "GET") {
    serverUrl += `?${Object.entries(requestBody)
      .filter(([, value]) => value !== undefined)
      .map(
        ([key, value]) => `${key}=${encodeURIComponent(JSON.stringify(value))}`
      )
      .join("&")}`;
  } else if (method.requestType === "POST") {
    requestOptions.body = JSON.stringify(requestBody);
  }

  try {
    const response = await fetch(serverUrl, {
      ...requestOptions,
      method: method.requestType,
    });

    if (response.ok) {
      const responseObj = decode<Response>(await response.text());

      if (method.onSuccess) {
        await dispatchTask(coreEngine, method.onSuccess);
      }

      await dispatchTask(coreEngine, responseObj.methods);
    } else if (method.onError) {
      await dispatchTask(coreEngine, method.onError);
    }
  } catch (e) {
    console.error(e);

    if (method.onError) {
      await dispatchTask(coreEngine, method.onError);
    }
  }

  if (method.after) {
    await dispatchTask(coreEngine, method.after);
  }
}

async function registerElement(
  element: Element,
  coreEngine: CoreEngine
): Promise<void> {
  const childElements: Element[] = [];
  const taskPromises: Promise<void>[] = [];

  for (const [key, value] of Object.entries(element)) {
    if (value instanceof Element) {
      taskPromises.push(
        coreEngine.taskQueue
          .run(registerElement, value, coreEngine)
          .then(() => {
            element[key] = getSimplifiedElement(value);

            childElements.push(element[key]);
          })
      );
    }

    if (
      _.isArray(value) &&
      _.every(value, (nestedElement) => nestedElement instanceof Element)
    ) {
      taskPromises.push(
        Promise.all(
          value.map((nestedElement) =>
            coreEngine.taskQueue.run(registerElement, nestedElement, coreEngine)
          )
        ).then(() => {
          element[key] = value.map((nestedElement) =>
            getSimplifiedElement(nestedElement)
          );
          childElements.push(...element[key]);
        })
      );
    }
  }

  await Promise.all(taskPromises);

  coreEngine.addToDispatchQueue(
    actions.setElement({
      element,
      childIds: childElements.map((nestedElement) => nestedElement.id),
    })
  );
}

async function dispatchTask(coreEngine: CoreEngine, methods?: Method[]) {
  if (_.isEmpty(methods)) {
    return;
  }

  await coreEngine.dispatch(async (__, getState) => {
    if (
      coreEngine.ownerId !== selectors.getCurrentTemplateOwnerId(getState())
    ) {
      return;
    }

    try {
      await Promise.all(
        methods!.map((method) => evalMethod(method, coreEngine))
      );
    } catch (e) {
      console.error("Fail to evalMethod", e);
    }

    try {
      coreEngine.flushDispatchQueue();
    } catch (e) {
      console.error("Fail to dispatch method", e);
    }
  });
}

export async function engineDispatch<T>(
  dispatch: ThunkDispatch<RootState, unknown, AnyAction>,
  methods?: Method[],
  clientInfo?: ClientInfo<T>
): Promise<void> {
  if (_.isEmpty(methods)) {
    return;
  }

  await dispatch(async (__, getState) => {
    const coreEngine = Builder(CoreEngine)
      .ownerId(selectors.getCurrentTemplateOwnerId(getState()))
      .taskQueue(getTaskQueue())
      .dispatch(dispatch)
      .clientInfo(clientInfo)
      .build();

    return dispatchTask(coreEngine, methods);
  });
}
