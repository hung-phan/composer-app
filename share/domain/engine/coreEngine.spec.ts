import {
  Button,
  Image,
  InputState,
  Layout,
  Text,
} from "../../elements/components/widgets";
import { makeStore } from "../../store";
import {
  Placeholder,
  RenderElementMethod,
  UpdateElementMethod,
  UpdateInListElementMethod,
} from "../interfaces";
import { engineDispatch } from "./coreEngine";
import { ROOT_ID } from "./index";

describe("coreEngine", () => {
  test("should update the store correctly", async () => {
    const store = makeStore({});

    await engineDispatch(store.dispatch, [
      RenderElementMethod.builder()
        .element(
          Layout.builder()
            .id(ROOT_ID)
            .items([
              Text.builder().id("text_element").build(),
              Image.builder()
                .id("image_element")
                .src("https://google.com")
                .build(),
              Button.builder().id("button_element").build(),
              Placeholder.builder().id("placeholder_element").build(),
              InputState.builder()
                .id("stateholder_element")
                .value("string_data")
                .build(),
            ])
            .build()
        )
        .build(),
    ]);

    expect(store.getState()).toMatchSnapshot();
  });

  test("should support override element with RenderElementMethod", async () => {
    const store = makeStore({});

    await engineDispatch(store.dispatch, [
      RenderElementMethod.builder()
        .element(
          Layout.builder()
            .id(ROOT_ID)
            .items([
              Image.builder()
                .id("image_element")
                .src("https://google.com")
                .build(),
              Placeholder.builder().id("placeholder_element").build(),
            ])
            .build()
        )
        .build(),
    ]);

    await engineDispatch(store.dispatch, [
      RenderElementMethod.builder()
        .element(
          Layout.builder()
            .id("placeholder_element")
            .items([
              Text.builder().id("text_element").build(),
              Button.builder().id("button_element").build(),
            ])
            .build()
        )
        .build(),
    ]);

    expect(store.getState()).toMatchSnapshot();
  });

  test("should remove element with RenderElementMethod", async () => {
    const store = makeStore({});

    await engineDispatch(store.dispatch, [
      RenderElementMethod.builder()
        .element(
          Layout.builder()
            .id(ROOT_ID)
            .items([
              Image.builder()
                .id("image_element")
                .src("https://google.com")
                .build(),
              Layout.builder()
                .id("placeholder_element")
                .items([
                  Text.builder().id("text_element").build(),
                  Button.builder().id("button_element").build(),
                ])
                .build(),
            ])
            .build()
        )
        .build(),
    ]);

    await engineDispatch(store.dispatch, [
      RenderElementMethod.builder()
        .element(Placeholder.builder().id("placeholder_element").build())
        .build(),
    ]);

    expect(store.getState()).toMatchSnapshot();
  });

  test("should support list replacement with same id", async () => {
    const store = makeStore({});

    await engineDispatch(store.dispatch, [
      RenderElementMethod.builder()
        .element(
          Layout.builder()
            .id(ROOT_ID)
            .items([
              Text.builder().id("text_element_1").build(),
              Placeholder.builder().id("placeholder_element").build(),
            ])
            .build()
        )
        .build(),
    ]);

    await engineDispatch(store.dispatch, [
      UpdateInListElementMethod.builder()
        .id("placeholder_element")
        .elements([
          Text.builder().id("text_element_2").build(),
          Placeholder.builder().id("placeholder_element").build(),
        ])
        .build(),
    ]);

    expect(store.getState()).toMatchSnapshot();
  });

  test("should support nested element update with same id", async () => {
    const store = makeStore({});

    await engineDispatch(store.dispatch, [
      RenderElementMethod.builder()
        .element(
          Layout.builder()
            .id(ROOT_ID)
            .items([
              Image.builder().id("image_element_1").build(),
              Placeholder.builder().id("placeholder_element").build(),
            ])
            .build()
        )
        .build(),
    ]);

    await engineDispatch(store.dispatch, [
      UpdateElementMethod.builder()
        .id("placeholder_element")
        .element(
          Layout.builder()
            .id("nested_layout_element")
            .items([
              Image.builder().id("image_element_2").build(),
              Placeholder.builder().id("placeholder_element").build(),
            ])
            .build()
        )
        .build(),
    ]);

    expect(store.getState()).toMatchSnapshot();
  });
});
