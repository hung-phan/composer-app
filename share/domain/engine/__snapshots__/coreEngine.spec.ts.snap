// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`coreEngine should remove element with RenderElementMethod 1`] = `
{
  "templateEngine": {
    "@@ENGINE_ROOT_ID": Node {
      "childs": {
        "image_element": true,
        "placeholder_element": true,
      },
      "element": Layout {
        "id": "@@ENGINE_ROOT_ID",
        "interfaceName": "Layout",
        "items": [
          Element {
            "id": "image_element",
            "interfaceName": "Image",
            Symbol(immer-draftable): true,
          },
          Element {
            "id": "placeholder_element",
            "interfaceName": "Placeholder",
            Symbol(immer-draftable): true,
          },
        ],
        Symbol(immer-draftable): true,
      },
      "interfaceName": "Node",
      Symbol(immer-draftable): true,
    },
    "image_element": Node {
      "element": Image {
        "id": "image_element",
        "interfaceName": "Image",
        "src": "https://google.com",
        Symbol(immer-draftable): true,
      },
      "interfaceName": "Node",
      "parent": "@@ENGINE_ROOT_ID",
      Symbol(immer-draftable): true,
    },
    "placeholder_element": Node {
      "element": Placeholder {
        "id": "placeholder_element",
        "interfaceName": "Placeholder",
        Symbol(immer-draftable): true,
      },
      "interfaceName": "Node",
      "parent": "@@ENGINE_ROOT_ID",
      Symbol(immer-draftable): true,
    },
  },
}
`;

exports[`coreEngine should support list replacement with same id 1`] = `
{
  "templateEngine": {
    "@@ENGINE_ROOT_ID": Node {
      "childs": {
        "placeholder_element": true,
        "text_element_1": true,
        "text_element_2": true,
      },
      "element": Layout {
        "id": "@@ENGINE_ROOT_ID",
        "interfaceName": "Layout",
        "items": [
          Element {
            "id": "text_element_1",
            "interfaceName": "Text",
            Symbol(immer-draftable): true,
          },
          Element {
            "id": "text_element_2",
            "interfaceName": "Text",
            Symbol(immer-draftable): true,
          },
          Element {
            "id": "placeholder_element",
            "interfaceName": "Placeholder",
            Symbol(immer-draftable): true,
          },
        ],
        Symbol(immer-draftable): true,
      },
      "interfaceName": "Node",
      Symbol(immer-draftable): true,
    },
    "placeholder_element": Node {
      "element": Placeholder {
        "id": "placeholder_element",
        "interfaceName": "Placeholder",
        Symbol(immer-draftable): true,
      },
      "interfaceName": "Node",
      "parent": "@@ENGINE_ROOT_ID",
      Symbol(immer-draftable): true,
    },
    "text_element_1": Node {
      "element": Text {
        "id": "text_element_1",
        "interfaceName": "Text",
        Symbol(immer-draftable): true,
      },
      "interfaceName": "Node",
      "parent": "@@ENGINE_ROOT_ID",
      Symbol(immer-draftable): true,
    },
    "text_element_2": Node {
      "element": Text {
        "id": "text_element_2",
        "interfaceName": "Text",
        Symbol(immer-draftable): true,
      },
      "interfaceName": "Node",
      "parent": "@@ENGINE_ROOT_ID",
      Symbol(immer-draftable): true,
    },
  },
}
`;

exports[`coreEngine should support nested element update with same id 1`] = `
{
  "templateEngine": {
    "@@ENGINE_ROOT_ID": Node {
      "childs": {
        "image_element_1": true,
        "nested_layout_element": true,
      },
      "element": Layout {
        "id": "@@ENGINE_ROOT_ID",
        "interfaceName": "Layout",
        "items": [
          Element {
            "id": "image_element_1",
            "interfaceName": "Image",
            Symbol(immer-draftable): true,
          },
          Element {
            "id": "nested_layout_element",
            "interfaceName": "Layout",
            Symbol(immer-draftable): true,
          },
        ],
        Symbol(immer-draftable): true,
      },
      "interfaceName": "Node",
      Symbol(immer-draftable): true,
    },
    "image_element_1": Node {
      "element": Image {
        "id": "image_element_1",
        "interfaceName": "Image",
        Symbol(immer-draftable): true,
      },
      "interfaceName": "Node",
      "parent": "@@ENGINE_ROOT_ID",
      Symbol(immer-draftable): true,
    },
    "image_element_2": Node {
      "element": Image {
        "id": "image_element_2",
        "interfaceName": "Image",
        Symbol(immer-draftable): true,
      },
      "interfaceName": "Node",
      "parent": "nested_layout_element",
      Symbol(immer-draftable): true,
    },
    "nested_layout_element": Node {
      "childs": {
        "image_element_2": true,
        "placeholder_element": true,
      },
      "element": Layout {
        "id": "nested_layout_element",
        "interfaceName": "Layout",
        "items": [
          Element {
            "id": "image_element_2",
            "interfaceName": "Image",
            Symbol(immer-draftable): true,
          },
          Element {
            "id": "placeholder_element",
            "interfaceName": "Placeholder",
            Symbol(immer-draftable): true,
          },
        ],
        Symbol(immer-draftable): true,
      },
      "interfaceName": "Node",
      "parent": "@@ENGINE_ROOT_ID",
      Symbol(immer-draftable): true,
    },
    "placeholder_element": Node {
      "element": Placeholder {
        "id": "placeholder_element",
        "interfaceName": "Placeholder",
        Symbol(immer-draftable): true,
      },
      "interfaceName": "Node",
      "parent": "nested_layout_element",
      Symbol(immer-draftable): true,
    },
  },
}
`;

exports[`coreEngine should support override element with RenderElementMethod 1`] = `
{
  "templateEngine": {
    "@@ENGINE_ROOT_ID": Node {
      "childs": {
        "image_element": true,
        "placeholder_element": true,
      },
      "element": Layout {
        "id": "@@ENGINE_ROOT_ID",
        "interfaceName": "Layout",
        "items": [
          Element {
            "id": "image_element",
            "interfaceName": "Image",
            Symbol(immer-draftable): true,
          },
          Element {
            "id": "placeholder_element",
            "interfaceName": "Layout",
            Symbol(immer-draftable): true,
          },
        ],
        Symbol(immer-draftable): true,
      },
      "interfaceName": "Node",
      Symbol(immer-draftable): true,
    },
    "button_element": Node {
      "element": Button {
        "id": "button_element",
        "interfaceName": "Button",
        "size": "lg",
        "style": "default",
        Symbol(immer-draftable): true,
      },
      "interfaceName": "Node",
      "parent": "placeholder_element",
      Symbol(immer-draftable): true,
    },
    "image_element": Node {
      "element": Image {
        "id": "image_element",
        "interfaceName": "Image",
        "src": "https://google.com",
        Symbol(immer-draftable): true,
      },
      "interfaceName": "Node",
      "parent": "@@ENGINE_ROOT_ID",
      Symbol(immer-draftable): true,
    },
    "placeholder_element": Node {
      "childs": {
        "button_element": true,
        "text_element": true,
      },
      "element": Layout {
        "id": "placeholder_element",
        "interfaceName": "Layout",
        "items": [
          Element {
            "id": "text_element",
            "interfaceName": "Text",
            Symbol(immer-draftable): true,
          },
          Element {
            "id": "button_element",
            "interfaceName": "Button",
            Symbol(immer-draftable): true,
          },
        ],
        Symbol(immer-draftable): true,
      },
      "interfaceName": "Node",
      "parent": "@@ENGINE_ROOT_ID",
      Symbol(immer-draftable): true,
    },
    "text_element": Node {
      "element": Text {
        "id": "text_element",
        "interfaceName": "Text",
        Symbol(immer-draftable): true,
      },
      "interfaceName": "Node",
      "parent": "placeholder_element",
      Symbol(immer-draftable): true,
    },
  },
}
`;

exports[`coreEngine should update the store correctly 1`] = `
{
  "templateEngine": {
    "@@ENGINE_ROOT_ID": Node {
      "childs": {
        "button_element": true,
        "image_element": true,
        "placeholder_element": true,
        "stateholder_element": true,
        "text_element": true,
      },
      "element": Layout {
        "id": "@@ENGINE_ROOT_ID",
        "interfaceName": "Layout",
        "items": [
          Element {
            "id": "text_element",
            "interfaceName": "Text",
            Symbol(immer-draftable): true,
          },
          Element {
            "id": "image_element",
            "interfaceName": "Image",
            Symbol(immer-draftable): true,
          },
          Element {
            "id": "button_element",
            "interfaceName": "Button",
            Symbol(immer-draftable): true,
          },
          Element {
            "id": "placeholder_element",
            "interfaceName": "Placeholder",
            Symbol(immer-draftable): true,
          },
          Element {
            "id": "stateholder_element",
            "interfaceName": "InputState",
            Symbol(immer-draftable): true,
          },
        ],
        Symbol(immer-draftable): true,
      },
      "interfaceName": "Node",
      Symbol(immer-draftable): true,
    },
    "button_element": Node {
      "element": Button {
        "id": "button_element",
        "interfaceName": "Button",
        "size": "lg",
        "style": "default",
        Symbol(immer-draftable): true,
      },
      "interfaceName": "Node",
      "parent": "@@ENGINE_ROOT_ID",
      Symbol(immer-draftable): true,
    },
    "image_element": Node {
      "element": Image {
        "id": "image_element",
        "interfaceName": "Image",
        "src": "https://google.com",
        Symbol(immer-draftable): true,
      },
      "interfaceName": "Node",
      "parent": "@@ENGINE_ROOT_ID",
      Symbol(immer-draftable): true,
    },
    "placeholder_element": Node {
      "element": Placeholder {
        "id": "placeholder_element",
        "interfaceName": "Placeholder",
        Symbol(immer-draftable): true,
      },
      "interfaceName": "Node",
      "parent": "@@ENGINE_ROOT_ID",
      Symbol(immer-draftable): true,
    },
    "stateholder_element": Node {
      "element": InputState {
        "id": "stateholder_element",
        "interfaceName": "InputState",
        "value": "string_data",
        Symbol(immer-draftable): true,
      },
      "interfaceName": "Node",
      "parent": "@@ENGINE_ROOT_ID",
      Symbol(immer-draftable): true,
    },
    "text_element": Node {
      "element": Text {
        "id": "text_element",
        "interfaceName": "Text",
        Symbol(immer-draftable): true,
      },
      "interfaceName": "Node",
      "parent": "@@ENGINE_ROOT_ID",
      Symbol(immer-draftable): true,
    },
  },
}
`;
